Water dipole analysis

lucasdekam opened the pull request:
Hello, I've implemented a class for analyzing the orientation of water molecules. It's based on the one in [WatAnalysis](https://github.com/ChiahsinChu/WatAnalysis) but a bit more flexible. The goal is to get a plot like this, where positive cos theta implies dipoles pointing to the right and negative cos theta dipoles pointing to the left.

![ori](https://github.com/user-attachments/assets/abd4d76c-4011-4fee-80fb-e8c916632535)

You can use the code as follows. First you identify the surfaces, like in the AtomDensity class:

```
from ase.io import read
from ectoolkits.structures.slab import Slab

xyz_file = ... # your trajectory xyz file, I tested using https://ai2db.ikkem.com/electroface/IF-100035

slab = Slab(read(xyz_file, index='0'))

surf_idx_up = slab.find_surf_idx('Pt', check_cross_boundary=True, dsur='up', tolerance=2)
surf_idx_dw = slab.find_surf_idx('Pt', check_cross_boundary=True, dsur='dw', tolerance=2)
```

Then you can perform water orientation analysis using the new code:

```
from ectoolkits.analysis.water import WaterOrientation

analysis = WaterOrientation(
    xyz=xyz_file,
    cell=slab.cell,
    surf1=surf_idx_up,
    surf2=surf_idx_dw,
    strict=False,
    origin="surf1",
)

analysis.run(step=10)  # step allows to skip some frames to speed up the calculation
```

Then plot with:

```
import matplotlib.pyplot as plt 
%matplotlib widget 

fig = plt.figure()
ax = fig.add_subplot()
analysis.plot_density(ax, sym=False)
```

or 

```
analysis.plot_orientation(ax, sym=False)
```

Let me know what you think, I'm open to any changes. Some things can still be improved, like combining the `density` function in `water.py` with the unit conversion in `atom_density.py`. For now I put all the changes in one file to avoid breaking something.

robinzyb commented:
wow super thanks! I will review this later. I am considerting refactoring my code recently because this package was written many years ago to provide better API and integrate different functions.

robinzyb commented:
could you add simple unit tests covering the main functions to ensure further improvement? 
Here is the `https://github.com/robinzyb/ECToolkits/tree/devel/tests/analysis/atom_density`

lucasdekam commented:
Sure, I'll set up some testing code in the coming week :-)

lucasdekam commented:
I've added some basic tests for the water density and orientation profiles, using the existing P-water and SnO2-water test trajectories. While testing I found some bugs and fixed them. I'm not very experienced with pytest so I'm open to any feedback (also on the other code). Thanks!

lucasdekam commented:
Looks like a typing issue that is not compatible with older Python versions, will fix

lucasdekam commented on ectoolkits/analysis/water.py:
```diff
@@ -0,0 +1,482 @@
+        # Setup Universe
+        universe = Universe(
+            xyz,
+            transformations=trans.boxdimensions.set_dimensions(cell.cellpar()),
```
actually seeing this again it might be good to make it work also for binary files, like XTC

robinzyb commented:
I merged this PR now. I will refactor together later on.

lucasdekam commented:
Cool, looking forward!
